packages/jyve-pyodide-unsafe/src/kernel.ts

```diff
@@ -0,0 +1,121 @@
+import {Kernel, KernelMessage} from '@jupyterlab/services';
+
+import {JSUnsafeKernel} from '@deathbeds/jyve-js-unsafe';
+import {JyveKernel} from '@deathbeds/jyve/lib/kernel';
+
+const {jyve} = (require('../package.json') as any);
+
+export const kernelSpec: Kernel.ISpecModel = jyve.kernelspec;
+
+// const DEBUG = false;
+
+
+export class PyodideUnsafeKernel extends JSUnsafeKernel {
+  protected kernelSpec = kernelSpec;
+  protected _pyodide: any;
+
+  jyveInfo(): KernelMessage.IInfoReply {
+    const jsInfo = super.jyveInfo();
+    return {
+      ...jsInfo,
+      help_links: [...jsInfo.help_links, ...jyve.help_links],
+      implementation: kernelSpec.name,
+      language_info: jyve.language_info
+    };
+  }
+
+  async pyodide() {
+    if (!this._pyodide) {
+      this._pyodide = await PyodideUnsafeKernel.pyodide(
+        (await this.iframe()).contentWindow
+      );
+    }
+    return this._pyodide;
+  }
+
+  resetUserNS() {
+    // TODO: Clear the namespace on the Python side
+    this.userNS = {};
+    this._pyodide = null;
+  }
+
+  async execNS(parent: KernelMessage.IMessage) {
+    const k = this;
+
+    let execNS = await super.execNS(parent);
+    const pyodide = await this.pyodide();
+    // pyodide.$options = {debug: DEBUG ? 10 : 0};
+
+    execNS = {
+      ...execNS,
+      __PYODIDE__: pyodide,
+    };
+
+    pyodide.write = (data: any) => {
+        k.sendJSON(k.fakeDisplayData(parent, {
+            'text/plain': `${data}`}));
+    };
+
+    pyodide.runPython(
+        "from js import window as _window\n" +
+        "import sys\n" +
+        "sys.stdout.write = _window.pyodide.write\n" +
+        "sys.stderr.write = _window.pyodide.write\n");
+
+    return execNS;
+  }
+
+  async transpile(code: string) {
+    await this.pyodide();
+    const src = JSON.stringify(code);
+
+    return `__PYODIDE__.runPython(${src});`;
+  }
+}
+
+function bootstrap_pyodide(window: any) {
+  let baseURL = "http://iodide-project.github.io/pyodide-demo/";
+  let wasmURL = `${baseURL}pyodide.asm.wasm?x=${Date.now()}`;
```

Comment:
In the near term, I'd love to see this hoisted as high as it can be (maybe even to package.json) and made a permalink backed by CDN, e.g.
https://cdn.rawgit.com/iodide-project/pyodide-demo/a39e3f3e80a438a07e6a8028fa4745768c16cbc2/pyodide.asm.wasm

Longer term, of course, i think we'd have to bite the bullet and figure out a way to actually
- ship that and its big brother `.data` in an npm package
- not have that get bundled into the main jupyterlab webpack bundle

One of the really nice things when demoing jyve is that going to https://deathbeds.github.io/jyve has pretty much always Just Worked... and since the Jyve can make the browser a touch... ahem... unstable, I end up having to reload a lot. The every-reload cache busting is a bummer in that way.

I think at some point, we (as in, the jupyterlab ecosystem) will need to seriously look at serviceworker/appcache approaches for some of these things.

Oh, and I guess I'd like it to work in chromium-based stuff. :blush:

Reply:
Interesting.  Didn't know about `cdn.rawgit.com`.  That's certainly better than the reload-forcing trick.  That seems like an easy improvement to do now.

I haven't spent any time thinking about distribution and deployment beyond that.  It's definitely suboptimal now.